### DISCUSSION TITLE
SignalR MaximumParallelInvocationsPerClient only run on one thread

### DISCUSSION THREAD

**Discussion by quytrieu37** [dotnet/aspnetcore#54835] [Q&A]
I'm currently using one HubConnection to execute multiple tasks on a hub. I've configured MaximumParallelInvocationsPerClient = 4 in the hub settings. It seems that when tasks complete quickly, they run concurrently. However, if there is a long-running task, it blocks other tasks from running.
To investigate this issue, I logged the thread IDs and found that all tasks from one connection use the same thread. Therefore, when a long-running task occurs, it blocks other tasks.
 I utilized Thread.Sleep to simulate a long-running task. 

![image](https://github.com/dotnet/aspnetcore/assets/74439440/ddb5e9dd-64ce-4337-a245-3c942feb1d2f)

![image](https://github.com/dotnet/aspnetcore/assets/74439440/bc622500-5d3c-4f2f-b35d-be94d9b2d14e)

As observed in the provided screenshot, all tasks are executed on the same thread. Consequently, when a long-running task occurs, it causes other tasks to be blocked.
![image](https://github.com/dotnet/aspnetcore/assets/74439440/8c3d528c-1df5-4695-bf44-35e20d0f154c)

To address this issue, I used Task.Run to execute my long-running task in the background, and everything worked fine.
![image](https://github.com/dotnet/aspnetcore/assets/74439440/7c7052ac-448e-4483-bc10-c24ad3af55cd)

![image](https://github.com/dotnet/aspnetcore/assets/74439440/238cead5-99ee-4237-9ddd-abff9a777d68)

However, I'm wondering if there is a configuration option to run concurrent tasks from one connection in multiple threads
And is the MaximumParallelInvocationsPerClient only allow each client call multiple hub method at same time but in the same thread?
Thank you!

**Comment by gfoidl** [comment]
Let's continue in https://github.com/dotnet/aspnetcore/issues/54836
